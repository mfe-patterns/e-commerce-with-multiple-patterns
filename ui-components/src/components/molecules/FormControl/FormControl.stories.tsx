import { storiesOf } from '@storybook/react';
import React from 'react';

import { FormControl } from './FormControl';
import { Wrapper } from './FormControl.styles';

storiesOf('Molecules', module).add('FormControl', () => (
  <>
    <Wrapper>
      <FormControl
        label="Label"
        placeholder="Placeholder"
        helpText="Filium morte multavit si sine dubio praeclara sunt"
      />
    </Wrapper>
    <Wrapper>
      <FormControl
        label="Label"
        placeholder="Placeholder"
        error
        helpText="Oops! Looks like you’re missing a few characters"
      />
    </Wrapper>
    <Wrapper>
      <FormControl
        label="Label"
        placeholder="Placeholder"
        success
        helpText="Way to go! You’ve completed your first task"
      />
    </Wrapper>
    <Wrapper>
      <FormControl
        label="Label"
        placeholder="Placeholder"
        disabled
        helpText="Fill the other field to enable this"
      />
    </Wrapper>
  </>
));
