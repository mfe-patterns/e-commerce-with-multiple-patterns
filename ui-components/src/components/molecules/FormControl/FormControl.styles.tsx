import styled from 'styled-components';
import { Label } from '../../atoms/Typography/Typography.styles';
import { HelpText } from '../../atoms/Typography/Typography';

export const Wrapper = styled.div`
  margin: 20px;
`;

export const LabelContainer = styled.div`
  flex-direction: row;
  align-items: baseline;
`;

export const StyledLabel = styled(Label)`
  margin-bottom: ${({ theme }) => `${theme.gridSize}px`};
`;

export const LabelHelpText = styled(HelpText)`
  margin-left: ${({ theme }) => `${theme.gridSize}px`};
`;

export const HelpTextContainer = styled.div`
  flex-direction: row;
  margin-top: ${({ theme }) => `${theme.gridSize}px`};
`;

export const HelpTextLeft = styled.div`
  flex: 1;
  margin-right: ${({ theme }) => `${theme.gridSize}px`};
`;
