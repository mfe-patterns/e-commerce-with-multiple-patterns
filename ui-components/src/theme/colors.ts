interface Shade {
  50: string;
  100: string;
  200: string;
  300: string;
  400: string;
  500: string;
  600: string;
  700: string;
}

const dark: Shade = {
  50: "#11111193",
  100: "#111111a2",
  200: "#111111B3",
  300: "#111111B9",
  400: "#111111CC",
  500: "#111111d8",
  600: "#111111eb",
  700: "#111111",
};
const neutral: Shade = {
  50: "#fafafa",
  100: "#f2f2f2",
  200: "#ededed",
  300: "#dedede",
  400: "#cccccc",
  500: "#666666",
  600: "#434343",
  700: "#343434",
};

const primary: Shade = {
  50: "#e3f2fd",
  100: "#bbdefb",
  200: "#90caf9",
  300: "#64b5f6",
  400: "#42a5f5",
  500: "#2196f3",
  600: "#1e88e5",
  700: "#1976d2",
};

const white = "#ffffff";

export interface IColorPallete {
  neutral: Shade;
  dark: Shade;
  primary: Shade;
  white: string;
  black: string;
  error: string;
  success: string;
  text: string;
  transparent: string;
  link: string;
}

export const colors: IColorPallete = {
  neutral,
  dark,
  primary,
  white,
  black: "#000000",
  error: "#dc3545",
  success: "#28a745",
  text: dark[700],
  transparent: "transparent",
  link: "#0064cc",
};
