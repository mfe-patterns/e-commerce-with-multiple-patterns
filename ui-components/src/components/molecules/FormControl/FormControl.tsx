import React from 'react';
import { HelpText, HelpTextType } from '../../atoms/Typography/Typography';
import type { ITextInputProps } from '../../atoms/TextInput/TextInput';
import { TextInput } from '../../atoms/TextInput/TextInput';
import * as S from './FormControl.styles';

interface FormControlProps extends ITextInputProps {
  label: string;
  error?: boolean;
  success?: boolean;
  helpText?: string;
}

export const FormControl = React.forwardRef<HTMLElement, FormControlProps>(
  (props: FormControlProps, ref) => {
    const {
      label,
      error,
      success,
      helpText,
      ...textInputProps
    } = props;

    return (
      <>
        <S.LabelContainer>
          <S.StyledLabel>
            {label}
          </S.StyledLabel>
        </S.LabelContainer>
        <TextInput
          ref={ref}
          error={error}
          success={success}
          {...textInputProps}
        />
        {helpText && (
          <S.HelpTextContainer>
            <S.HelpTextLeft>
              <HelpText
                type={
                  error
                    ? HelpTextType.ERROR
                    : success
                      ? HelpTextType.SUCCESS
                      : HelpTextType.DEFAULT
                }>
                {helpText}
              </HelpText>
            </S.HelpTextLeft>
          </S.HelpTextContainer>
        )}
      </>
    );
  },
);
