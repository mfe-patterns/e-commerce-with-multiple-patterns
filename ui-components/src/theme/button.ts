import { colors } from "./colors";

interface IButtonType {
  color: string;
  backgroundColor?: string;
  borderColor?: string;
  fontFamily: string;
  border: string;
  disabled: {
    color: string;
    backgroundColor?: string;
    borderColor?: string;
  };
  horizontalPadding: boolean;
}

export const primary: IButtonType = {
  color: colors.white,
  borderColor: colors.transparent,
  fontFamily: "Roboto-Regular",
  border: "none",
  disabled: {
    color: colors.white,
    borderColor: colors.transparent,
  },
  horizontalPadding: true,
};
export const secondary: IButtonType = {
  backgroundColor: colors.white,
  color: colors.text,
  fontFamily: "Roboto-Regular",
  border: "1px solid",
  disabled: {
    backgroundColor: colors.white,
    color: colors.dark[100],
  },
  horizontalPadding: true,
};

interface IButtonSize {
  height: string;
  borderRadius: string;
  fontSize: string;
  lineHeight: string;
  letterSpacing: string;
  paddingVertical: string;
  paddingHorizontal: string;
}

interface ISize {
  xlarge: IButtonSize;
  large: IButtonSize;
  medium: IButtonSize;
  small: IButtonSize;
  xsmall: IButtonSize;
}

const size: ISize = {
  xlarge: {
    height: "56px",
    borderRadius: "28px",
    fontSize: "16px",
    lineHeight: "24px",
    letterSpacing: "0.7px",
    paddingVertical: "15px",
    paddingHorizontal: "29.5px",
  },
  large: {
    height: "46px",
    borderRadius: "23px",
    fontSize: "15px",
    lineHeight: "20px",
    letterSpacing: "0.7px",
    paddingVertical: "12px",
    paddingHorizontal: "25.5px",
  },
  medium: {
    height: "40px",
    borderRadius: "20px",
    fontSize: "14px",
    lineHeight: "20px",
    letterSpacing: "0.7px",
    paddingVertical: "10px",
    paddingHorizontal: "21.5px",
  },
  small: {
    height: "36px",
    borderRadius: "18px",
    fontSize: "13px",
    lineHeight: "19px",
    letterSpacing: "0.7px",
    paddingVertical: "8px",
    paddingHorizontal: "20px",
  },
  xsmall: {
    height: "29px",
    borderRadius: "14.5px",
    fontSize: "12px",
    lineHeight: "19px",
    letterSpacing: "0.7px",
    paddingVertical: "5px",
    paddingHorizontal: "14px",
  },
};

export interface IButton {
  type: {
    primary: IButtonType;
    secondary: IButtonType;
  },
  size: ISize;
}

export const button: IButton = {
  type: {
    primary: {
      ...primary,
      disabled: {
        ...primary.disabled,
        backgroundColor: colors.primary[200],
      },
      backgroundColor: colors.primary[500],
    },
    secondary: {
      ...secondary,
      borderColor: colors.primary[500],
      disabled: {
        ...secondary.disabled,
        borderColor: colors.primary[200],
      },
    },
  },
  size,
};
