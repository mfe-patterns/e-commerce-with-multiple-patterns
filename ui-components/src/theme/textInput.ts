import { colors } from './colors';
import { FontFamily } from './typography';

interface ITextInputType {
  color: string;
  placeholderColor?: string;
  backgroundColor: string;
  borderColor: string;
}

export interface ITextInput {
  fontFamily: FontFamily;
  default: ITextInputType;
  disabled: ITextInputType;
  error: ITextInputType;
  success: ITextInputType;
}

export const textInput: ITextInput = {
  fontFamily: 'Roboto-Regular',
  default: {
    color: colors.text,
    placeholderColor: colors.dark[100],
    backgroundColor: colors.white,
    borderColor: colors.neutral[300],
  },
  disabled: {
    color: colors.dark[50],
    backgroundColor: colors.neutral[50],
    borderColor: colors.neutral[300],
  },
  error: {
    color: colors.text,
    backgroundColor: colors.white,
    borderColor: colors.error,
  },
  success: {
    color: colors.text,
    backgroundColor: colors.white,
    borderColor: colors.success,
  },
};
