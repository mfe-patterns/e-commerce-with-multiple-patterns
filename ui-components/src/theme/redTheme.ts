import { ITypography, typography } from "./typography";
import { IColorPallete, colors } from "./colors";
import { ITextInput, textInput } from "./textInput";
import { button, IButton, primary, secondary } from "./button";

const redPalette: IColorPallete = {
  ...colors,
  primary: {
    50: "#ffebee",
    100: "#ffcdd2",
    200: "#ef9a9a",
    300: "#e57373",
    400: "#ef5350",
    500: "#f44336",
    600: "#e53935",
    700: "#d32f2f",
  },
};

const redButton: IButton = {
  ...button,
  type: {
    primary: {
      ...primary,
      disabled: {
        ...primary.disabled,
        backgroundColor: redPalette.primary[200],
      },
      backgroundColor: redPalette.primary[500],
    },
    secondary: {
      ...secondary,
      borderColor: redPalette.primary[500],
      disabled: {
        ...secondary.disabled,
        borderColor: redPalette.primary[200],
      },
    },
  },
};

interface ITheme {
  typography: ITypography;
  colors: IColorPallete;
  textInput: ITextInput;
  button: IButton;
  gridSize: number;
}

export const theme: ITheme = {
  typography,
  colors: redPalette,
  textInput,
  button: redButton,
  gridSize: 8,
};
