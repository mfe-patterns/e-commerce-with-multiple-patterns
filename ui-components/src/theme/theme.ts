import { ITypography, typography } from "./typography";
import { IColorPallete, colors } from "./colors";
import { ITextInput, textInput } from "./textInput";
import { button, IButton } from "./button";

interface ITheme {
  typography: ITypography;
  colors: IColorPallete;
  textInput: ITextInput;
  button: IButton;
  gridSize: number;
}

export const theme: ITheme = {
  typography,
  colors,
  textInput,
  button,
  gridSize: 8
};
