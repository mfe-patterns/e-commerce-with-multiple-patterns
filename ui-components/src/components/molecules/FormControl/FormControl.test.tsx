import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';

import { FormControl } from './FormControl';
import { theme } from '../../../theme/theme';

describe('FormControl', () => {
  it('should match the snapshot', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <FormControl
          label="Text Input"
          helpText="Filium morte multavit si sine dubio praeclara sunt"
          data-testid="text-input"
        />
        ,
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot for error', () => {
    const { container, getByText } = render(
      <ThemeProvider theme={theme}>
        <FormControl
          error
          label="Text Input"
          helpText="Filium morte multavit si sine dubio praeclara sunt"
          data-testid="text-input"
        />
        ,
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
    expect(getByText('Filium morte multavit si sine dubio praeclara sunt')).toHaveStyleRule(
      'color',
      '#dc3545'
    );
  });

  it('should match the snapshot for success', () => {
    const { container, getByText } = render(
      <ThemeProvider theme={theme}>
        <FormControl
          success
          label="Text Input"
          helpText="Filium morte multavit si sine dubio praeclara sunt"
          data-testid="text-input"
        />
        ,
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
    expect(getByText('Filium morte multavit si sine dubio praeclara sunt')).toHaveStyleRule(
      'color',
      '#28a745'
    );
  });
});
