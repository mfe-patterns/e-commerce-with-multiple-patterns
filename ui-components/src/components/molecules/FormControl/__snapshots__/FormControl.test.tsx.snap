// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormControl should match the snapshot 1`] = `
.c1 {
  font-family: Roboto-Regular;
  font-size: 14px;
  line-height: 21px;
  -webkit-letter-spacing: 0.7px;
  -moz-letter-spacing: 0.7px;
  -ms-letter-spacing: 0.7px;
  letter-spacing: 0.7px;
  color: #111111;
}

.c7 {
  font-family: Roboto-Regular;
  font-size: 11px;
  line-height: 16px;
  -webkit-letter-spacing: 0.7px;
  -moz-letter-spacing: 0.7px;
  -ms-letter-spacing: 0.7px;
  letter-spacing: 0.7px;
  color: #111111CC;
}

.c4 {
  height: 48px;
  padding: 12px 16px;
  color: #111111;
  background-color: #ffffff;
  border: 1px solid;
  border-color: #dedede;
  box-sizing: border-box;
  width: 100%;
}

.c3 {
  min-height: 48px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

.c0 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c2 {
  margin-bottom: 8px;
}

.c5 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-top: 8px;
}

.c6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-right: 8px;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1 c2"
    >
      Text Input
    </label>
  </div>
  <div
    class="c3"
  >
    <input
      class="c4"
      data-testid="text-input"
    />
  </div>
  <div
    class="c5"
  >
    <div
      class="c6"
    >
      <p
        class="c7"
        color="#111111CC"
      >
        Filium morte multavit si sine dubio praeclara sunt
      </p>
    </div>
  </div>
  ,
</div>
`;

exports[`FormControl should match the snapshot for error 1`] = `
.c1 {
  font-family: Roboto-Regular;
  font-size: 14px;
  line-height: 21px;
  -webkit-letter-spacing: 0.7px;
  -moz-letter-spacing: 0.7px;
  -ms-letter-spacing: 0.7px;
  letter-spacing: 0.7px;
  color: #111111;
}

.c7 {
  font-family: Roboto-Regular;
  font-size: 11px;
  line-height: 16px;
  -webkit-letter-spacing: 0.7px;
  -moz-letter-spacing: 0.7px;
  -ms-letter-spacing: 0.7px;
  letter-spacing: 0.7px;
  color: #dc3545;
}

.c4 {
  height: 48px;
  padding: 12px 16px;
  color: #111111;
  background-color: #ffffff;
  border: 1px solid;
  border-color: #dedede;
  box-sizing: border-box;
  width: 100%;
  color: #111111;
  background-color: #ffffff;
  border-color: #dc3545;
  padding-right: 48px;
}

.c3 {
  min-height: 48px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

.c0 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c2 {
  margin-bottom: 8px;
}

.c5 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-top: 8px;
}

.c6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-right: 8px;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1 c2"
    >
      Text Input
    </label>
  </div>
  <div
    class="c3"
  >
    <input
      class="c4"
      data-testid="text-input"
    />
  </div>
  <div
    class="c5"
  >
    <div
      class="c6"
    >
      <p
        class="c7"
        color="#dc3545"
      >
        Filium morte multavit si sine dubio praeclara sunt
      </p>
    </div>
  </div>
  ,
</div>
`;

exports[`FormControl should match the snapshot for success 1`] = `
.c1 {
  font-family: Roboto-Regular;
  font-size: 14px;
  line-height: 21px;
  -webkit-letter-spacing: 0.7px;
  -moz-letter-spacing: 0.7px;
  -ms-letter-spacing: 0.7px;
  letter-spacing: 0.7px;
  color: #111111;
}

.c7 {
  font-family: Roboto-Regular;
  font-size: 11px;
  line-height: 16px;
  -webkit-letter-spacing: 0.7px;
  -moz-letter-spacing: 0.7px;
  -ms-letter-spacing: 0.7px;
  letter-spacing: 0.7px;
  color: #28a745;
}

.c4 {
  height: 48px;
  padding: 12px 16px;
  color: #111111;
  background-color: #ffffff;
  border: 1px solid;
  border-color: #dedede;
  box-sizing: border-box;
  width: 100%;
  color: #111111;
  background-color: #ffffff;
  border-color: #28a745;
  padding-right: 48px;
}

.c3 {
  min-height: 48px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

.c0 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c2 {
  margin-bottom: 8px;
}

.c5 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-top: 8px;
}

.c6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-right: 8px;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1 c2"
    >
      Text Input
    </label>
  </div>
  <div
    class="c3"
  >
    <input
      class="c4"
      data-testid="text-input"
    />
  </div>
  <div
    class="c5"
  >
    <div
      class="c6"
    >
      <p
        class="c7"
        color="#28a745"
      >
        Filium morte multavit si sine dubio praeclara sunt
      </p>
    </div>
  </div>
  ,
</div>
`;
